src/Rings/oscar_singular.jl

```diff
@@ -0,0 +1,448 @@
+
+# prime field, small characteristic
+iso_oscar_singular_coeff_ring(R::fpField) = OscarSingularCoefficientRingMapGeneric(R, Singular.Fp(Int(characteristic(R))))
+
+# ZZ/nZZ, n small and big
+struct OscarSingularCoefficientRingMapZZMod{D} <: OscarSingularCoefficientRingMap{D, Singular.N_ZnRing}
+  R::D
+  S::Singular.N_ZnRing
+end
+
+iso_oscar_singular_coeff_ring(R::Union{zzModRing, ZZModRing}) = OscarSingularCoefficientRingMapZZMod(R, Singular.residue_ring(Singular.Integers(), BigInt(modulus(R)))[1]
+)
+
+# QQ(a)
+struct OscarSingularCoefficientRingMapAbsSimpleNumField <: OscarSingularCoefficientRingMap{AbsSimpleNumField, Singular.N_AlgExtField}
+  R::AbsSimpleNumField
+  S::Singular.N_AlgExtField
+end
+
+function iso_oscar_singular_coeff_ring(R::AbsSimpleNumField)
+  minpoly = defining_polynomial(R)
+  Qa = parent(minpoly)
+  a = gen(Qa)
+  SQa, (Sa,) = Singular.FunctionField(Singular.QQ, _variables_for_singular(symbols(Qa)))
+  Sminpoly = SQa(coeff(minpoly, 0))
+  for i in 1:degree(minpoly)
+    Sminpoly += SQa(coeff(minpoly, i))*Sa^i
+  end
+  SK, _ = Singular.AlgebraicExtensionField(SQa, Sminpoly)
+  return OscarSingularCoefficientRingMapAbsSimpleNumField(R, SK)
+end
+
+# It is nonsense, but the conversion resides in Singular.jl/src/number/n_algExt.jl
+# via parent object call overloading
+
+# GF(p, n), small p
+
+struct OscarSingularCoefficientRingMapfqPolyRepField <: OscarSingularCoefficientRingMap{fqPolyRepField, Singular.N_ZpField}
+  R::fqPolyRepField
+  S::Singular.N_AlgExtField
+end
+
+function iso_oscar_singular_coeff_ring(F::fqPolyRepField)
+  # TODO: the Fp(Int(char)) can throw
+  minpoly = modulus(F)
+  Fa = parent(minpoly)
+  SFa, (Sa,) = Singular.FunctionField(Singular.Fp(Int(characteristic(F))),
+                                                    _variables_for_singular(symbols(Fa)))
+  Sminpoly = SFa(coeff(minpoly, 0))
+  for i in 1:degree(minpoly)
+    Sminpoly += SFa(coeff(minpoly, i))*Sa^i
+  end
+  SF, _ = Singular.AlgebraicExtensionField(SFa, Sminpoly)
+  return OscarSingularCoefficientRingMapfqPolyRepField(F, SF)
+end
+
+# Finite field (FqField)
+struct OscarSingularCoefficientRingMapFqField{D} <: OscarSingularCoefficientRingMap{FqField, D}
+  R::FqField
+  S::D
+  iso::MapFromFunc{FqField, FqField}
+
+  OscarSingularCoefficientRingMapFqField(R, S) = new{typeof(S)}(R, S)
+  OscarSingularCoefficientRingMapFqField(R, S, iso) = new{typeof(S)}(R, S, iso)
+end
+
+function _absolute_field(F::FqField)
+  if is_absolute(F)
+    error("don't use me, you are already absolute")
+  end
+  Kx = parent(defining_polynomial(F))
+  Fabs, QQtoFabs = Nemo._residue_field(defining_polynomial(F); absolute = true, check = false)
+  Fabsx, = polynomial_ring(Fabs, :x; cached = false)
+  return Fabs, MapFromFunc(Fabs, F, a -> F(preimage(QQtoFabs, a)), b -> begin a = Fabs(); Nemo.set!(a, b); a end)
+end
+
+# Nonsense for FqField (aka fq_default from flint)
```

Comment:
What is nonsense about it? If it is nonsense why is it here? Maybe it does have a sense after all? Or maybe the comment is the nonsense? Argh, help, I can't make sense of this :-(